usr/share/rear/layout/save/GNU/Linux/210_raid_layout.sh

```diff
@@ -1,109 +1,223 @@
+            component_device=$( get_device_name $component_device )
+            test -b "$component_device" || Error "RAID $raiddevice component device '$component_device' is no block device"
+            # A component device that is more than one word would make 'read' fail for this 'raid' entry in disklayout.conf
+            test $component_device || Error "RAID $raiddevice component device '$component_device' is not a single word"
+            # Have the component devices string as "first_component_device,second_component_device,..."
+            test $component_devices && component_devices+=",$component_device" || component_devices="$component_device"
+        done
+    fi
+    test $component_devices && raid_layout_entry+=" devices=$component_devices" || Error "No component devices for RAID $raiddevice"
+
+    # Optional mdadm option settings:
+    test $name && raid_layout_entry+=" name=$name"
+
+    line=( $( grep "Version :" $mdadm_details ) )
+    metadata=${line[2]}
+    test $metadata && raid_layout_entry+=" metadata=$metadata"
+
+    line=( $( grep "UUID :" $mdadm_details ) )
+    uuid=${line[2]}
+    test $uuid && raid_layout_entry+=" uuid=$uuid"
+
+    line=( $( grep "Layout :" $mdadm_details ) )
+    layout=${line[2]}
+    # fix up layout for RAID10:
+    # > near=2,far=1 -> n2
+    if test "$level" = "raid10" ; then
+        OIFS=$IFS
+        IFS=","
+        for param in "$layout" ; do
+            copies=${layout%=*}
+            number=${layout#*=}
+            test "$number" -gt 1 && layout="${copies:0:1}$number"
+        done
+        IFS=$OIFS
+    fi
+    # mdadm can print '-unknown-' for a RAID layout
+    # which got recently (2019-12-02) added to RAID0 (it existed before for RAID5 and RAID6 and RAID10) see
+    # https://git.kernel.org/pub/scm/utils/mdadm/mdadm.git/commit/Detail.c?id=329dfc28debb58ffe7bd1967cea00fc583139aca
+    # so we treat '-unknown-' same as an empty value to avoid that layout/prepare/GNU/Linux/120_include_raid_code.sh
+    # will create a 'mdadm' command in diskrestore.sh like "mdadm ... --layout=-unknown- ..." which would fail
+    # during "rear recover" with something like "mdadm: layout -unknown- not understood for raid0"
+    # see https://github.com/rear/rear/issues/2616
+    test $layout -a '-unknown-' != "$layout" && raid_layout_entry+=" layout=$layout"
+
+    chunksize=$( grep "Chunk Size" $mdadm_details | tr -d " " | cut -d ":" -f "2" | sed -r 's/^([0-9]+).+/\1/')
+    test $chunksize && raid_layout_entry+=" chunk=$chunksize"
+
+    spare_devices=""
+    test $raid_devices && spare_devices=$(( total_devices - raid_devices ))
+    test $spare_devices -gt 0 && raid_layout_entry+=" spare-devices=$spare_devices"
+
+    test $container_size -gt 0 && raid_layout_entry+=" size=$container_size"
+
+    echo "$raid_layout_entry" >>$DISKLAYOUT_FILE
+
+    extract_partitions "$raiddevice" >>$DISKLAYOUT_FILE
+
+done < <( mdadm --detail --scan --config=partitions )
```

Comment:
I wonder why not simply prepend the loop with `mdadm --detail --scan --config=partitions | ` instead of using this more complicated and less standard (bash-specific) idiom? I also believe the version with a pipe will be more readable, because the data source will come first and thew destination next. I see that the idiom `loop < <(command)` is being used in a huge number of places in ReaR, and until now I have never asked why.

Reply:
You are right.
This is a leftover of bad old code style.
For reasoning why `COMMAND | while read ... do ... done`
is usually better than `while read ... do ... done < <( COMMAND )` see
https://github.com/rear/rear/blob/master/usr/share/rear/layout/save/GNU/Linux/220_lvm_layout.sh#L25

Reply:
Very interesting reference, thanks. I have not realized that "the drawback of using a pipe is that the "while read ... do ... done" part is run as separated process (in a subshell) so that e.g. one cannot set variables in the "while read ... do ... done" part that are meant to be used after the pipe." which in many cases may be the reason for having to use the old code style.

Reply:
Implemented via
https://github.com/rear/rear/pull/2714/commits/e980b65154c67b82bb69b1ca874fbbb01e102dc1

We have to carefully test it because now
extract_partitions is run as a separated process (in a subshell)
because it runs in the "mdadm ... | while read ... ; do ... done" pipe
so things only work if extract_partitions does not set variables
that are meant to be used outside of this pipe, check extract_partitions()
in layout/save/GNU/Linux/200_partition_layout.sh